python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_handler.py

```diff
@@ -0,0 +1,509 @@
+        self._otel_span.set_attribute(key, value)
+
+    def record_exception(self, exception: BaseException) -> None:
+        self._otel_span.record_exception(exception)
+
+    def end(self, status: trace_api.Status) -> None:
+        self[OPENINFERENCE_SPAN_KIND] = self._span_kind or CHAIN
+        self._otel_span.set_status(status=status)
+        self._otel_span.end()
+
+    @property
+    def context(self) -> context_api.Context:
+        return trace_api.set_span_in_context(self._otel_span)
+
+    @singledispatchmethod
+    def process_event(self, event: BaseEvent) -> None:
+        logger.warning(f"Unhandled event of type {event.__class__.__qualname__}")
+
+    @process_event.register
+    def _(self, event: AgentChatWithStepStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = AGENT
+        self[INPUT_VALUE] = event.user_msg
+
+    @process_event.register
+    def _(self, event: AgentChatWithStepEndEvent) -> None:
+        self[OUTPUT_VALUE] = str(event.response)
+
+    @process_event.register
+    def _(self, event: AgentRunStepStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = AGENT
+        if input := event.input:
+            self[INPUT_VALUE] = input
+
+    @process_event.register
+    def _(self, event: AgentRunStepEndEvent) -> None:
+        # FIXME: not sure what to do here
+        ...
+
+    @process_event.register
+    def _(self, event: AgentToolCallEvent) -> None:
+        tool = event.tool
+        if name := tool.name:
+            self[TOOL_NAME] = name
+        self[TOOL_DESCRIPTION] = tool.description
+        self[TOOL_PARAMETERS] = json.dumps(tool.get_parameters_dict())
+
+    @process_event.register
+    def _(self, event: StreamChatErrorEvent) -> None:
+        self.record_exception(event.exception)
+
+    @process_event.register
+    def _(self, event: StreamChatDeltaReceivedEvent) -> None:
+        if self._first_token_timestamp is None:
+            self._first_token_timestamp = datetime.now(timezone.utc)
+
+    @process_event.register
+    def _(self, event: EmbeddingStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = EMBEDDING
+
+    @process_event.register
+    def _(self, event: EmbeddingEndEvent) -> None:
+        for i, (text, vector) in enumerate(zip(event.chunks, event.embeddings)):
+            self[f"{EMBEDDING_EMBEDDINGS}.{i}.{EMBEDDING_TEXT}"] = text
+            self[f"{EMBEDDING_EMBEDDINGS}.{i}.{EMBEDDING_VECTOR}"] = vector
+
+    @process_event.register
+    def _(self, event: LLMPredictStartEvent) -> None:
+        if not self._span_kind:
+            self._span_kind = LLM
+        self[LLM_PROMPT_TEMPLATE] = event.template.get_template()
+        vars = event.template.template_vars
+        args = {**event.template.kwargs, **(event.template_args if event.template_args else {})}
+        template_args = {var: arg for var in vars if (arg := args.get(var)) is not None}
```

Comment:
```suggestion
        variable_names = event.template.template_vars
        args = {**event.template.kwargs, **(event.template_args if event.template_args else {})}
        template_variables = {var_name: var_value for var_name in variable_names if (var_value := args.get(var_name)) is not None}
```

Suggestion: More verbose names would help here.

Reply:
updated